### ISSUE TITLE
arm64 build not working on M1 or Raspberry Pi 4 64-bit

### ISSUE THREAD

**Issue by alanivey** [abtreece/confd#14]
I'm having trouble with `arm64` on a Raspberry Pi 4 on the 64-bit OS, and a colleague with an M1 Mac running confd in a docker container had the same issue:

```bash
$ apt update

$ apt install -y --no-install-recommends ca-certificates wget

$ export CONFD_VERSION=0.18.0
$ export dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')"

$ wget -O /usr/local/bin/confd "https://github.com/abtreece/confd/releases/download/v${CONFD_VERSION}/confd-${CONFD_VERSION}-linux-${dpkgArch}"

$ confd -version
confd 0.18.0 (Git SHA: d0061cc, Go Version: go1.15.8)

$ confd
Illegal instruction (core dumped)
```

I am able to get things working by building it on the Pi via `docker run --rm -it debian bash`:

```bash
$ apt update

$ apt install -y git make wget

$ wget https://golang.org/dl/go1.15.8.linux-arm64.tar.gz

$ tar -C /usr/local -x -z -f go1.15.8.linux-arm64.tar.gz

$ export PATH="$PATH:/usr/local/go/bin"

$ go version
go version go1.15.8 linux/arm64

$ git clone https://github.com/abtreece/confd.git "$HOME/confd-src"

$ cd "$HOME/confd-src"

$ git checkout v0.18.0

$ make build

$ bin/confd -version
confd 0.18.0 (Git SHA: d0061cc, Go Version: go1.15.8)

$ bin/confd
2021-05-05T14:37:45Z 8be84966b714 bin/confd[7498]: INFO Backend set to etcd
2021-05-05T14:37:45Z 8be84966b714 bin/confd[7498]: INFO Starting confd
2021-05-05T14:37:45Z 8be84966b714 bin/confd[7498]: INFO Backend source(s) set to http://127.0.0.1:2379
2021-05-05T14:37:45Z 8be84966b714 bin/confd[7498]: WARNING Cannot load template resources: confdir '/etc/confd' does not exist
```

For what it's worth, I launched a c6g instance in AWS and *both* methods worked successfully. I don't know enough about the differences in the CPU architecture between Graviton2 vs M1 vs Pi4 to know why, but hopefully this information helps. I did not take my compiled binary from my Pi4 and try it on the AWS instance though...

**Comment by abtreece** [comment]
@alanivey to prove it's `upx` causing the issue I've built an arm64 binary. Can you try it on your rpi4 and confirm if it works? I ordered a rpi4 as well so I can tinker with builds on that hardware.
[confd-0.18.1-dev-linux-arm64.gz](https://github.com/abtreece/confd/files/6442759/confd-0.18.1-dev-linux-arm64.gz)

**Comment by abtreece** [comment]
I have an rpi4 now does seem to be a `upx` issue.

```
abtreece@rpi4:~$ ./confd-0.18.0-linux-arm64 --version
Illegal instruction (core dumped)

abtreece@rpi4:~$ ./confd-0.18.1-dev-linux-arm64 --version
confd 0.18.1-dev (Git SHA: 79bf98b, Go Version: go1.15.8)
```

**Comment by abtreece** [comment]
@alanivey... It appears to be an issue with both golang 1.15.8 **and** upx. I switched to compiling the arm64 version using upx 3.94 and it still failed with "Illegal instruction". 

Building with golang 1.15.12 produces a binary that works on aarch64. It appears some compiler issues were fixed in [1.15.10](https://github.com/golang/go/issues?q=milestone%3AGo1.15.10+label%3ACherryPickApproved) and [1.15.11](https://github.com/golang/go/issues?q=milestone%3AGo1.15.11+label%3ACherryPickApproved).

Here is a build for arm64 compiled with 1.15.12...
[confd-0.18.1-dev-linux-arm64.gz](https://github.com/abtreece/confd/files/6448529/confd-0.18.1-dev-linux-arm64.gz)

I am going to test the other builds to confirm nothing else is borked and will likely release a 0.18.1

cc: https://github.com/Islandora-Devops/isle-buildkit/issues/117

**Comment by alanivey** [comment]
@abtreece thank you for looking into this!

On my Pi4 in a Debian container it now works as expected:

```bash
$ ./confd-0.18.1-dev-linux-arm64 --version
confd 0.18.1-dev (Git SHA: 79bf98b, Go Version: go1.15.12)

$  ./confd-0.18.1-dev-linux-arm64
2021-05-10T12:27:30Z 7438df406abe ./confd-0.18.1-dev-linux-arm64[2764]: INFO Backend set to etcd
2021-05-10T12:27:30Z 7438df406abe ./confd-0.18.1-dev-linux-arm64[2764]: INFO Starting confd
2021-05-10T12:27:30Z 7438df406abe ./confd-0.18.1-dev-linux-arm64[2764]: INFO Backend source(s) set to http://127.0.0.1:2379
2021-05-10T12:27:30Z 7438df406abe ./confd-0.18.1-dev-linux-arm64[2764]: WARNING Cannot load template resources: confdir '/etc/confd' does not exist
```

I'll edit this post after I have a colleague test on an M1 Mac.

**Comment by abtreece** [comment]
@alanivey I too wish I had an M1... I'm looking at using `goreleaser` for more robust and consistent builds. I've attached a darwin-arm64 build for your colleague to try. 🤞 

[confd_v0.18.0-next_Darwin_ARM64.tar.gz](https://github.com/abtreece/confd/files/6475998/confd_v0.18.0-next_Darwin_ARM64.tar.gz)

**Comment by alanivey** [comment]
@abtreece that works on the macOS shell. There's still something about the linux-arm64 binary not working in Docker though; not sure if that can be further troubleshooted if neither of us have one.

**Comment by abtreece** [comment]
@alanivey I might have acquired an M1 mini on eBay and it might be arriving tomorrow... help is on the way (hopefully).

**Comment by abtreece** [comment]
@alanivey it appears the issue running confd on Debian running via Docker on an M1 Mac is due to UPX. :smh:

At this point, I think the use of UPX has caused more problems than actual value... 

> What about UPX? If you’re not familiar with UPX, it’s an amazing tool that reduces the size of binaries. It does so by compressing the binary, and adding a small stub to uncompress and run it transparently. If we want to reduce the footprint of our containers, UPX will also be very counter-productive. First, the disk and network usage won’t be reduced a single bit, since layers are compressed anyway; so UPX won’t get us anything here.
>
> When running a normal binary, it is mapped in memory, so that only the bits that are needed get loaded (or “paged in”) when necessary. When running a binary compressed with UPX, the entire binary has to be uncompressed in memory. This results in higher memory usage and longer start times, especially with runtimes like Go that tend to generate bigger binaries.

So, I am going stop using UPX and switch the build/release process to GoReleaser. The resulting binaries will be archived and compressed for download but not themselves compressed. I'll try to have new binaries out in the next couple days

([quote ref](https://www.ardanlabs.com/blog/2020/04/docker-images-part3-going-farther-reduce-image-size.html))

**Comment by alanivey** [comment]
Sounds great, thanks for finding the solution!

**Comment by abtreece** [comment]
@alanivey I released [v0.18.1](https://github.com/abtreece/confd/releases/tag/v0.18.1) which I hope resolves the assorted issues related to UPX. I've done some simple testing on amd64 and arm64 macs, arm64 debian in docker on arm64 mac, rpi4 (arm64), rpi3 (armv7)... all of them printed a version and started. 🙏🏻 

```
./confd --version
confd 0.18.1 (Git SHA: 943899f, Go Version: go1.16.4)
```
